### PR TITLE
ruby: Encode decode cleanup and behavior normalization

### PR DESCRIPTION
**Pull request by skippy**
Hey @cfallin 

This PR may be a bridge too far.  I am sending this PR as *DO NOT MERGE*; if you agree with the direction of the changs, I'll add some tests (in the old-style format) and documentation.  

This pr started as just wanting to add common ruby syntactic sugar, specifically `#to_json` and `#to_proto`.  Then I realized that the jruby `.encode_json` was broken ([gist](https://gist.github.com/skippy/89cafb280017d52a582c)).  But it also had the very nice and useful `#to_h` which the c-ext did not.

I tried consolidating the json encoding/decoding into pure ruby, using the stdlib JSON module, the JSON gem (using the c-extension), and Yajl.  This ripped out a lot of c code, and a bit of java code, and I liked the simplification.  Until I ran benchmarks.  tl;dr: the native json library (haberman/upb) in this lib was consistantly 10x faster ([gist](https://gist.github.com/skippy/93d6c51947960260ab63)).  Yowzers....

I don't want to reopen the conversation in #121, and definitely not the tenor.  However, I agree with the concerns that having two fully separate implementations is going to be hard to maintain.  I'm finding more inconsitancies than I expected, and I'm just starting to scratch the surface.  (e.g., another is that the jruby extension doesn't check STRING encoding during init. like the c-ext does).  *BUT* it is alpha :)

If the JSON performance was closer, I would have submitted the PR with consolidated logic and have tried to make the case that moving some code into ruby is a good thing from a consistancy and maintenance perspective.  However, the performance difference is so great I don't feel I should make that argument.

So here is a less-aggressive PR with fewer changes.  Again, if this is moving in an acceptable direction, I'll finish it up.

### CODE REVIEW

File: ruby/ext/google/protobuf_c/message.c [hunk 1]
```diff
@@ -329,6 +329,30 @@ VALUE Message_inspect(VALUE _self) {
   return str;
 }
 
+
+VALUE Message_to_h(VALUE _self) {
+  MessageHeader* self;
+  TypedData_Get_Struct(_self, MessageHeader, &Message_type, self);
+
+  VALUE hash = rb_hash_new();
+  rb_gc_register_address(&hash);
```
Comment (skippy) [inline ruby/ext/google/protobuf_c/message.c]: is this needed if I use `rb_hash_new()`?  I sort of suspect not

File: ruby/ext/google/protobuf_c/message.c [hunk 2]
```diff
@@ -329,6 +329,30 @@ VALUE Message_inspect(VALUE _self) {
...
+
+  upb_msg_field_iter it;
+  for (upb_msg_field_begin(&it, self->descriptor->msgdef);
+       !upb_msg_field_done(&it);
+       upb_msg_field_next(&it)) {
+    const upb_fielddef* field = upb_msg_iter_field(&it);
+    VALUE msg_value = layout_get(self->descriptor->layout, Message_data(self), field);
+    VALUE msg_key   = rb_str_new2(upb_fielddef_name(field));
```
Comment (skippy) [inline ruby/ext/google/protobuf_c/message.c]: I need to double check, but I think rails moved its 'active_record' methods to return symbols as the keys.  I would like to use symbols as they are more efficient and the rails change may give cover to make that change...?  But it would be a breaking change, so I'm leaving the old behavior for now (though the `#new` method requires symbols...)

File: ruby/ext/google/protobuf_c/message.c [hunk 3]
```diff
@@ -407,6 +435,8 @@ VALUE build_class_from_descriptor(Descriptor* desc) {
   rb_define_method(klass, "clone", Message_dup, 0);
   rb_define_method(klass, "==", Message_eq, 1);
   rb_define_method(klass, "hash", Message_hash, 0);
+  rb_define_method(klass, "to_h", Message_to_h, 0);
+  rb_define_method(klass, "to_hash", Message_to_h, 0);
```
Comment (skippy) [inline ruby/ext/google/protobuf_c/message.c]: `#to_hash` is a common alias

File: ruby/lib/google/protobuf/message_exts.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        # def decode_json(json)
+        #   opts = {
+        #     symbolize_names: true,
+        #     symbolize_keys: true   #for Yajl
+
+        #   }
+        #   hash = JSON.parse(json, opts)
+        #   self.new(hash)
+        # end
+      end
```
Comment (skippy) [inline ruby/lib/google/protobuf/message_exts.rb]: (I'll clean this up if the PR direction is worth going in)

File: ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java [hunk 5]
```diff
@@ -346,6 +346,8 @@ public IRubyObject toHash(ThreadContext context) {
             IRubyObject value = getField(context, fdef);
             if (value.respondsTo("to_h")) {
                 value = Helpers.invoke(context, value, "to_h");
+            } else if (value.respondsTo("to_a")) {
+                value = Helpers.invoke(context, value, "to_a");
```
Comment (skippy) [inline ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java]: see the gist for the invalid jruby json encoding

**Comment by skippy** [comment]
your call, but:
* having a pure ruby method of `#to_h` that is used by both implementations is 'better' (hard time defining that... more maintainable?  I'm not so sure... clearer?  yes.  dry-er?  yes...but does that really matter?
* but raw-c is 9x faster ([gist](https://gist.github.com/skippy/63c9025db92019886aa0))

but damn, if the java and jruby versions were a bit closer in performance, then the argument for moving most of the logic to ruby with a little java in the performant spots, and then a bit more of the C code could be made.

**Comment by haberman** [comment]
> Until I ran benchmarks. tl;dr: the native json library (haberman/upb) in this lib was consistantly 10x faster (gist). Yowzers....

I would also mention that the upb JSON encoder is not very optimized (it uses `snprintf()` to format numbers, for example). Once this is optimized, the performance difference will probably be even greater than 10x.

By the way, if you are discovering inconsistencies between implementations, *please* add what you discovered to the unit tests. Not only does this make everything more robust, it makes refactorings like this much more tractable and lower-risk.

File: ruby/ext/google/protobuf_c/message.c [hunk 6]
```diff
@@ -329,6 +329,31 @@ VALUE Message_inspect(VALUE _self) {
...
+  VALUE hash = rb_hash_new();
+  rb_gc_register_address(&hash);
+
+  upb_msg_field_iter it;
+  for (upb_msg_field_begin(&it, self->descriptor->msgdef);
+       !upb_msg_field_done(&it);
+       upb_msg_field_next(&it)) {
+    const upb_fielddef* field = upb_msg_iter_field(&it);
+    VALUE msg_value = layout_get(self->descriptor->layout, Message_data(self), field);
+    VALUE msg_key   = ID2SYM(upb_fielddef_name(field));
```
Comment (skippy) [inline ruby/ext/google/protobuf_c/message.c]: **NOTE** this makes the hash return symbols for keys.  I think this is the right call for 2 reasons:
* hell of a lot faster
* primary reason: the message initializer only accepts symbolized keys

File: ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java [hunk 7]
```diff
@@ -346,8 +346,10 @@ public IRubyObject toHash(ThreadContext context) {
...
             }
-            ret.fastASet(runtime.newString(fdef.getName()), value);
+            ret.fastASet(runtime.newSymbol(fdef.getName()), value);
```
Comment (skippy) [inline ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java]: **NOTE** this is a breaking change.  It brings it inline with the initializers, and it is quite a bit faster, but it is a breaking change at that...

**Comment by skippy** [comment]
@haberman I am adding tests; on every one that I find and fix.  Yet I'm eager to dive into tests a bit more as there are a bunch of edge cases that aren't being tested.  I hope to start get to that later this week.

nice work on haberman/upb btw.

**Comment by cfallin** [comment]
Hi @skippy,

Thanks for all the work here! I'll give a closer look tomorrow. But one high-level request, especially since the flurry of pull-requests and comments is becoming somewhat hectic: could you put together a design/summary document of sorts listing (i) the discrepancies you're finding between the MRI Ruby extension and JRuby extension, and (ii) the specific semantic and API changes you're proposing? It would be easier to read that first and then review the code as appropriate. A shared Google doc is fine, or a gist in Markdown, or whatnot.

It's been a while since I reviewed the JRuby extension and wrote the MRI one so the summary is useful for me personally, at least :-)

Thanks!

**Comment by haberman** [comment]
@skippy Thanks for adding the tests and thanks for your nice words about upb!

**Comment by skippy** [comment]
@cfallin of course; thank you for asking.  An initial draft is here:  https://gist.github.com/skippy/00c23e4e7bdac5124fe4.  

No rush on my end!

**Comment by cfallin** [comment]
Thanks for that -- I'll try to take a look in the next few days!

File: ruby/ext/google/protobuf_c/message.c (hunk 1 above)
Comment (cfallin) [inline ruby/ext/google/protobuf_c/message.c]: Oh, definitely not! `rb_gc_register_address` registers a static root with the garbage collector -- e.g., for a global variable. You're telling the GC that some random address on this thread's stack is a root to be kept alive. Probably not what you want :-)

File: ruby/lib/google/protobuf/message_exts.rb (hunk 4 above)
Comment (cfallin) [inline ruby/lib/google/protobuf/message_exts.rb]: Given the performance delta, I'd rather keep JSON in C, though I wish it were faster to have it at this level, too.

File: ruby/ext/google/protobuf_c/message.c (hunk 1 above)
Comment (skippy) [inline ruby/ext/google/protobuf_c/message.c]: ha!  yes; sorry, I should have dived into the docs a bit more on that one.  thanks for the explanation.

File: ruby/lib/google/protobuf/message_exts.rb (hunk 4 above)
Comment (skippy) [inline ruby/lib/google/protobuf/message_exts.rb]: +1.  FYI: the `JSON` gem is using the c-ext when running under MRI and it is still quite a bit slower than the `upd` library.  kind of amazing actually

File: ruby/ext/google/protobuf_c/protobuf.c [hunk 8]
```diff
@@ -86,13 +86,6 @@ void Init_protobuf_c() {
   RepeatedField_register(protobuf);
   Map_register(protobuf);
 
-  rb_define_singleton_method(protobuf, "encode", Google_Protobuf_encode, 1);
-  rb_define_singleton_method(protobuf, "decode", Google_Protobuf_decode, 2);
-  rb_define_singleton_method(protobuf, "encode_json",
-                             Google_Protobuf_encode_json, 1);
-  rb_define_singleton_method(protobuf, "decode_json",
-                             Google_Protobuf_decode_json, 2);
-
```
Comment (skippy) [inline ruby/ext/google/protobuf_c/protobuf.c]: these didn't exist in the jruby version

File: ruby/lib/google/protobuf.rb [hunk 9]
```diff
@@ -36,3 +39,25 @@
...
+    def self.decode(klass, proto)
+      klass.decode(proto)
+    end
+
+    def self.decode_json(klass, json)
+      klass.decode_json(json)
+    end
+
+  end
+end
```
Comment (skippy) [inline ruby/lib/google/protobuf.rb]: these methods existed in both jruby and mri, but I think it is cleaner to just define them once

File: ruby/tests/basic.rb [hunk 10]
```diff
@@ -822,6 +822,67 @@ def test_parse_serialize
...
+      m = TestMessage.new(:optional_bool => true, :optional_double => -10.100001, :optional_string => 'foo', :repeated_string => ['bar1', 'bar2'])
+      expected_result = {
+        :optional_bool=>true,
+        :optional_bytes=>"",
+        :optional_double=>-10.100001,
+        :optional_enum=>:Default,
+        :optional_float=>0.0,
+        :optional_int32=>0,
+        :optional_int64=>0,
+        :optional_msg=>nil,
```
Comment (skippy) [inline ruby/tests/basic.rb]: the jruby version was returning a `{}`, but `nil`, if the msg is not set, seems to be the appropriate response

**Comment by skippy** [comment]
@cfallin what ya think?

**Comment by cfallin** [comment]
The unification effort in general is really helpful here, thanks!

The only potential concern I have is in adding `#to_proto` and `#to_json` methods to message objects. It's definitely more convenient than having to invoke `MyClass.encode(msg)`, but I recall a fairly spirited discussion around this when we designed the API and there were arguments to keep the message objects' method list "clean", i.e., only have overrides to `Object` methods. I realize the cat is partially out of the bag already with `#to_h`, and from my perspective today at least, the convenience methods *do* seem nice (no need to write out the message class's name explicitly; also allows for generic methods that take a message of any type).

@haberman, do you have any opinion here?

(To be clear, this is my only real potential concern, and I'm happy to take the rest of the changes!)

**Comment by skippy** [comment]
@cfallin very fair.  `#to_h` is a common extension and one worth keeping, esp. after Rails, which I think popularized that helper (though I'm not totally clear on the etymology).  `#to_json` has become more popular as a helper, usually used in the context of creating a data representation (such as a hash) and then, at the last minute, serializing it.  It is common pattern in the ruby on rails and api realms.  `#to_proto` is pure syntax sugar, and one that I liked from a few other ruby proto libraries.  

As an FYI, and why this pattern has emerged, is that frameworks like Rails, and other ruby gems, have a common short-hand where you can use syntax such as `render json: msg`, and it will call `msg.to_json`.  And that pattern holds for other helpers such as `render proto: msg` => `msg.to_proto`.  

The ruby community loves its syntax sugar, that is for sure.

Except for `#to_h`, which makes retrieving keys and values from the object very convenient, I won't push hard (or perhaps I should say, push my luck) for the other two. :)  Let me know what you all decide, and I'll go in and make the changes.

thanks!

**Comment by haberman** [comment]
The unfortunate thing about adding methods on the message objects is that is mixes the namespace of field names (from the .proto file) with the namespace of helper/convenience methods. `msg.to_json` is a helper method but `msg.to_jsonx` is a message field. This is really unclear to a reader. If a message field is named `to_json`, the user will need to use `msg["to_json"]` instead. If we add any convenience methods in the future, we could break existing code for any users who used that method name as a protobuf field name.

All that said, it's becoming apparent that the convenience and convention of helper methods has such strong precedent in Ruby culture that we'd be remiss not to support it. It's clear that we should support `#to_proto` and `#to_json`. Others we should evaluate on a case-by-case basis.

**Comment by cfallin** [comment]
OK, I'm in agreement too. Since we're still in alpha, I don't feel too bad about encroaching-by-eminent-domain into field namespace and grabbing a few method names. I agree that we should be careful about this in the future so that we don't break users (especially if we ever want to use a short/common/ambiguous name as a method).

The change itself looks good, so LGTM, and I'll merge!

**Comment by skippy** [comment]
@cfallin I found a bug mapping #to_h through the Map type.  I'm tracking down right now.  I'll beef up the test cases in that area as well.  It will take me, possibly, EOW to get that patch in.  

but shit.  sorry about that.

**Comment by cfallin** [comment]
@skippy OK, no worries, thanks for letting us know! If you'd like or if you need to pass it off, I'm happy to track it down given a repro, otherwise we're happy to take your fix.
